staging/src/k8s.io/client-go/tools/events/event_broadcaster.go

```diff
@@ -0,0 +1,290 @@
+
+			if isomorphicEvent.Series != nil {
+				isomorphicEvent.Series.Count++
+				isomorphicEvent.EventTime = metav1.MicroTime{Time: clock.Now()}
+
+			} else {
+				isomorphicEvent.Series = &v1beta1.EventSeries{
+					Count:            1,
+					LastObservedTime: metav1.MicroTime{Time: clock.Now()},
+				}
+				attemptRecording(eventBroadcaster.sink, isomorphicEvent, true, randGen, eventBroadcaster.sleepDuration)
+
+			}
+
+		} else {
+			eventBroadcaster.eventCache[eventKey] = event
+			attemptRecording(eventBroadcaster.sink, event, false, randGen, eventBroadcaster.sleepDuration)
+		}
+
+	}()
+
+}
+
+func attemptRecording(sink EventSink, event *v1beta1.Event, b bool, randGen *rand.Rand, sleepDuration time.Duration) {
+	tries := 0
+
+	for {
+		if recordEvent(sink, event, b) {
+			break
+		}
+		tries++
+		if tries >= maxTriesPerEvent {
+			klog.Errorf("Unable to write event '%#v' (retry limit exceeded!)", event)
+			break
+		}
+		// Randomize the first sleep so that various clients won't all be
+		// synced up if the master goes down.
+		if tries == 1 {
+			time.Sleep(time.Duration(float64(sleepDuration) * randGen.Float64()))
+		} else {
+			time.Sleep(sleepDuration)
+		}
+	}
+
+}
+
+func recordEvent(sink EventSink, event *v1beta1.Event, isEventSerie bool) bool {
+	var newEvent *v1beta1.Event
+	var err error
+	if isEventSerie {
+
+		patch, err := createPatchBytes(event)
+		if err != nil {
+			klog.Errorf("Unable to calculate diff, no merge is possible: %v", err)
+			return false
+		}
+		newEvent, err = sink.Patch(event, patch)
+	}
+	// Update can fail because the event may have been removed and it no longer exists.
+	if !isEventSerie || (isEventSerie && util.IsKeyNotFoundError(err)) {
+		// Making sure that ResourceVersion is empty on creation
+		event.ResourceVersion = ""
+		newEvent, err = sink.Create(event)
+	}
+	if err == nil {
+
+		event = newEvent
+		return true
+	}
+	// If we can't contact the server, then hold everything while we keep trying.
+	// Otherwise, something about the event is malformed and we should abandon it.
+	switch err.(type) {
+	case *restclient.RequestConstructionError:
+		// We will construct the request the same next time, so don't keep trying.
+		klog.Errorf("Unable to construct event '%#v': '%v' (will not retry!)", event, err)
+		return true
+	case *errors.StatusError:
+		if errors.IsAlreadyExists(err) {
+			klog.V(5).Infof("Server rejected event '%#v': '%v' (will not retry!)", event, err)
+		} else {
+			klog.Errorf("Server rejected event '%#v': '%v' (will not retry!)", event, err)
+		}
+		return true
+	case *errors.UnexpectedObjectError:
+		// We don't expect this; it implies the server's response didn't match a
+		// known pattern. Go ahead and retry.
+	default:
+		// This case includes actual http transport errors. Go ahead and retry.
+	}
+	klog.Errorf("Unable to write event: '%v' (may retry after sleeping)", err)
+	return false
+
+}
+
+func createPatchBytes(event *v1beta1.Event) ([]byte, error) {
```

Comment:
This name is misleading - it creates a patch between this event and event with series field clear, right?

Reply:
Yes